### PR TITLE
Create callgraph

### PR DESCRIPTION
**Pull request by Geoffrey1014**
add callgraph class

### CODE REVIEW

File: svf/include/Graphs/CallGraph.h [hunk 1]
```diff
@@ -0,0 +1,345 @@
...
+ */
+typedef GenericEdge<CallGraphNode> GenericCallGraphEdgeTy;
+class CallGraphEdge : public GenericCallGraphEdgeTy
+{
+
+public:
+    typedef Set<const CallICFGNode*> CallInstSet;
+    enum CEDGEK
+    {
+        CallRetEdge,TDForkEdge,TDJoinEdge,HareParForEdge
```
Comment (jumormt) [inline svf/include/Graphs/CallGraph.h:58]: Only callretedge

File: svf/include/Graphs/CallGraph.h [hunk 2]
```diff
@@ -0,0 +1,345 @@
...
+    {
+        return (cs << EdgeKindMaskBits) | k;
+    }
+    /// Get direct and indirect calls
+    //@{
+    inline CallSiteID getCallSiteID() const
+    {
+        return csId;
+    }
+    inline CallInstSet& getDirectCalls()
```
Comment (jumormt) [inline svf/include/Graphs/CallGraph.h:86]: This can be removed.

File: svf/include/Graphs/CallGraph.h [hunk 3]
```diff
@@ -0,0 +1,345 @@
...
+    }
+    inline const CallInstSet& getDirectCalls() const
+    {
+        return directCalls;
+    }
+    //@}
+
+    /// Add direct callsite
+    //@{
+    void addDirectCallSite(const CallICFGNode* call);
```
Comment (jumormt) [inline svf/include/Graphs/CallGraph.h:98]: can be made protected

File: svf/include/Graphs/CallGraph.h [hunk 4]
```diff
@@ -0,0 +1,345 @@
...
+        FunToCallGraphNodeMap::const_iterator it = funToCallGraphNodeMap.find(fun);
+        assert(it!=funToCallGraphNodeMap.end() && "call graph node not found!!");
+        return it->second;
+    }
+
+    //@}
+
+    /// Add/Get CallSiteID
+    //@{
+    inline CallSiteID addCallSite(const CallICFGNode* cs, const SVFFunction* callee)
```
Comment (jumormt) [inline svf/include/Graphs/CallGraph.h:279]: can be made private

File: svf/include/Graphs/PTACallGraph.h [hunk 5]
```diff
@@ -40,15 +40,16 @@ namespace SVF
...
 
 /*
  * Call Graph edge representing a calling relation between two functions
  * Multiple calls from function A to B are merged into one call edge
  * Each call edge has a set of direct callsites and a set of indirect callsites
  */
-typedef GenericEdge<PTACallGraphNode> GenericCallGraphEdgeTy;
-class PTACallGraphEdge : public GenericCallGraphEdgeTy
+typedef GenericEdge<PTACallGraphNode> GenericPTACallGraphEdgeTy;
+class PTACallGraphEdge : public GenericPTACallGraphEdgeTy
```
Comment (jumormt) [inline svf/include/Graphs/PTACallGraph.h:52]: make necessary changes similar to CallGraphEdge

File: svf/include/Graphs/PTACallGraph.h [hunk 6]
```diff
@@ -237,8 +238,8 @@ class PTACallGraphNode : public GenericCallGraphNodeTy
 /*!
  * Pointer Analysis Call Graph used internally for various pointer analysis
  */
-typedef GenericGraph<PTACallGraphNode, PTACallGraphEdge> GenericCallGraphTy;
-class PTACallGraph : public GenericCallGraphTy
+typedef GenericGraph<PTACallGraphNode, PTACallGraphEdge> GenericPTACallGraphTy;
+class PTACallGraph : public GenericPTACallGraphTy
```
Comment (jumormt) [inline svf/include/Graphs/PTACallGraph.h:242]: make necessary changes similar to CallGraph

File: svf/include/Graphs/CallGraph.h (hunk 3 above)
Comment (Geoffrey1014) [reply svf/include/Graphs/CallGraph.h:98]: callgraphedge's addDirectCallSite is used in CallGraph, so it cannot be protected

Comment (yuleisui) [reply svf/include/Graphs/CallGraph.h:98]: If it is only used in 'CallGraph', it can be made as private.

File: svf/include/Graphs/CallGraph.h [hunk 7]
```diff
@@ -0,0 +1,327 @@
...
+    typedef OrderedMap<const CallICFGNode*, FunctionSet> CallEdgeMap;
+
+    enum CGEK
+    {
+        NormCallGraph
+    };
+
+private:
+    /// Call site information
+    static CallSiteToIdMap csToIdMap;	///< Map a pair of call instruction and callee to a callsite ID
```
Comment (yuleisui) [inline svf/include/Graphs/CallGraph.h:218]: Do we need csToIdMap, idToCSMap and totalCallSiteNum in CallGraph?

**Comment by Geoffrey1014** [comment]
--- master.txt	2024-11-05 15:14:46
+++ callgraph.txt	2024-11-05 15:14:47
@@ -30,9 +30,9 @@
 VarArrayObj         831
 VarStructObj        1664
 ----------------Time and memory stats--------------------
-LLVMIRTime          4.101
-SVFIRTime           3.806
-SymbolTableTime     0.658
+LLVMIRTime          4.081
+SVFIRTime           2.555
+SymbolTableTime     0.517
 #######################################################
 
 *********PTACallGraph Stats (Andersen analysis)***************
\ No newline at end of file
@@ -59,11 +59,11 @@
 CollapseTime        0
 CopyGepTime         0
 LoadStoreTime       0
-MemoryUsageVmrss    4.16458e+06
-MemoryUsageVmsize   4.16454e+06
+MemoryUsageVmrss    4.16461e+06
+MemoryUsageVmsize   4.16455e+06
 SCCDetectTime       0
 SCCMergeTime        0
-TotalTime           121.433
+TotalTime           116.863
 UpdateCGTime        0
 ----------------Numbers stats----------------------------
 AddrProcessed       17887
\ No newline at end of file
@@ -114,47 +114,47 @@
 ****Persistent Points-To Cache Statistics: Andersen's analysis bitvector****
 ################ (program : redis-server.bc)###############
 UniquePointsToSets       44837
-TotalUnions              673052
+TotalUnions              673053
 PropertyUnions           222658
 UniqueUnions             19260
 LookupUnions             395148
-PreemptiveUnions         35986
+PreemptiveUnions         35987
 TotalComplements         3364152
 PropertyComplements      2998574
 UniqueComplements        16887
 LookupComplements        331806
 PreemptiveComplements    16885
-TotalIntersections       4182803
-PropertyIntersections    4145525
+TotalIntersections       4182804
+PropertyIntersections    4145533
 UniqueIntersections      842
-LookupIntersections      2147
-PreemptiveIntersections  34289
+LookupIntersections      2139
+PreemptiveIntersections  34290
 #######################################################
 
 *********Memory SSA Statistics***************
 ################ (program : redis-server.bc)###############
 ----------------Time and memory stats--------------------
-AverageRegSize      85.0301
-GenMUCHITime        1.86
-GenRegionTime       485.119
-InsertPHITime       0.571
-SSARenameTime       0.127
-TotalMSSATime       487.692
+AverageRegSize      85.727
+GenMUCHITime        1.791
+GenRegionTime       501.038
+InsertPHITime       0.503
+SSARenameTime       0.092
+TotalMSSATime       503.432
 ----------------Numbers stats----------------------------
 BBHasMSSAPhi        10755
-CSChiNode           112337
+CSChiNode           111246
 CSHasChi            22679
 CSHasMu             24549
-CSMuNode            164766
-FunEntryChi         24535
+CSMuNode            163193
+FunEntryChi         24294
 FunHasEntryChi      3967
 FunHasRetMu         4630
-FunRetMu            24466
+FunRetMu            24225
 LoadHasMu           24342
 LoadMuNode          43202
-MSSAPhi             60232
+MSSAPhi             59315
 MaxRegSize          1650
-MemRegions          4580
+MemRegions          4538
 StoreChiNode        14269
 StoreHasChi         10507
 #######################################################
\ No newline at end of file
@@ -162,16 +162,16 @@
 *********SVFG Statistics***************
 ################ (program : redis-server.bc)###############
 ----------------Time and memory stats--------------------
-ATNodeTime          0.915
-AvgWeight           166.005
+ATNodeTime          0.912
+AvgWeight           167.789
 ConnDirEdgeTime     0
-ConnIndEdgeTime     4.553
+ConnIndEdgeTime     4.525
 OptTime             0
 TLNodeTime          0
-TotalTime           5.468
+TotalTime           5.437
 ----------------Numbers stats----------------------------
-ActualIn            164766
-ActualOut           112337
+ActualIn            163193
+ActualOut           111246
 ActualParam         57426
 ActualRet           7530
 Addr                17887
\ No newline at end of file
@@ -183,25 +183,25 @@
 DirectCallEdge      41434
 DirectEdge          292884
 DirectRetEdge       7351
-FormalIn            24535
-FormalOut           24466
+FormalIn            24294
+FormalOut           24225
 FormalParam         7869
 FormalRet           795
 Gep                 84876
-IndCallEdge         192075
-IndRetEdge          126074
-IndirectEdge        724691
-IndirectEdgeLabels  120302478
+IndCallEdge         189998
+IndRetEdge          124671
+IndirectEdge        717043
+IndirectEdgeLabels  120311897
 Load                24342
-MSSAPhi             60232
+MSSAPhi             59315
 MaxInDegree         9710
 MaxIndInDeg         9710
 MaxIndOutDeg        2458
 MaxOutDegree        9600
 PHI                 6776
 Store               24164
-TotalEdge           1017575
-TotalNode           632140
+TotalEdge           1009927
+TotalNode           628077
 #######################################################
 
 *********PTACallGraph Stats (Flow-sensitive analysis)***************
\ No newline at end of file
@@ -222,28 +222,28 @@
 ----------------Time and memory stats--------------------
 AddrTime            0
 AverageSCCSize      19.105
-AvgPtsSize          67.7966
+AvgPtsSize          67.7948
 AvgTopLvlPtsSize    369.968
-AvgVersionPtsSize   32.3075
+AvgVersionPtsSize   32.3067
 CopyTime            0
 DirectPropaTime     0
 GepTime             0
 IndirectPropaTime   0
 LoadTime            0
-MemoryUsageVmrss    3.57764e+06
-MemoryUsageVmsize   3.61505e+06
+MemoryUsageVmrss    3.57665e+06
+MemoryUsageVmsize   3.61197e+06
 PhiTime             0
-PrelabelingTime     0.131
+PrelabelingTime     0.125
 ProcessTime         0
 PropagationTime     0
 SCCTime             0
-SolveTime           422.656
+SolveTime           406.923
 StoreTime           0
 Strong/WeakUpdTime  0
-TotalTime           637.671
+TotalTime           621.122
 UpdateCGTime        0
 VersionPropTime     0
-meldLabelingTime    211.668
+meldLabelingTime    211.009
 ----------------Numbers stats----------------------------
 CopysNum            14138
 DummyFieldPtrs      2305
\ No newline at end of file
@@ -261,33 +261,33 @@
 Pointers            494739
 ProcessedAParam     0
 ProcessedAddr       107322
-ProcessedCopy       162617
+ProcessedCopy       160966
 ProcessedFRet       0
-ProcessedGep        867729
-ProcessedLoad       899817
-ProcessedMSSANode   2318016
-ProcessedPhi        104061
-ProcessedStore      325348
+ProcessedGep        862436
+ProcessedLoad       840146
+ProcessedMSSANode   2293638
+ProcessedPhi        102466
+ProcessedStore      314283
 SolveIterations     6
 StoresNum           24164
 StrongUpdates       3378
 TotalEmptyVPts      0
-TotalExistingVPts   15080114
-TotalNonEmptyVPts   15080114
+TotalExistingVPts   15080539
+TotalNonEmptyVPts   15080539
 TotalObjects        15741
 TotalPointers       497044
 TotalSingleVObjs    0
-TotalVersions       17441804
+TotalVersions       17442525
 #######################################################
 
 ****Persistent Points-To Cache Statistics: flow-sensitive analysis bitvector****
 ################ (program : redis-server.bc)###############
-UniquePointsToSets       29667
-TotalUnions              1169886217
-PropertyUnions           1012294655
-UniqueUnions             38241
-LookupUnions             157501857
-PreemptiveUnions         51464
+UniquePointsToSets       29529
+TotalUnions              1100930485
+PropertyUnions           955694200
+UniqueUnions             37377
+LookupUnions             145148532
+PreemptiveUnions         50376
 TotalComplements         0
 PropertyComplements      0
 UniqueComplements        0
\ No newline at end of file

**Pull request merged**